engine/jobmaster/dm/dm_jobmaster.go

```diff
@@ -245,35 +249,28 @@ func (jm *JobMaster) OnMasterMessage(topic p2p.Topic, message interface{}) error
 
 // CloseImpl implements JobMasterImpl.CloseImpl
 func (jm *JobMaster) CloseImpl(ctx context.Context) error {
-	jm.Logger().Info("closing the dm jobmaster")
-	if err := jm.taskManager.OperateTask(ctx, dmpkg.Delete, nil, nil); err != nil {
-		return err
-	}
+	return jm.messageAgent.Close(ctx)
+}
 
-outer:
-	for {
-		select {
-		case <-ctx.Done():
-			return ctx.Err()
-		case <-time.After(time.Second):
-			// wait all worker offline
-			jm.workerManager.SetNextCheckTime(time.Now())
-			// manually call Tick since outer event loop is closed.
-			jm.workerManager.Tick(ctx)
-			if len(jm.workerManager.WorkerStatus()) == 0 {
-				if err := jm.checkpointAgent.Remove(ctx); err != nil {
-					jm.Logger().Error("failed to remove checkpoint", zap.Error(err))
-				}
-				break outer
-			}
-		}
-	}
+// OnCancel implements JobMasterImpl.OnCancel
+func (jm *JobMaster) OnCancel(ctx context.Context) error {
```

Comment:
JobMasterImpl.OnCancel and StopImpl is not found 🤔

Reply:
Yes, this method should be called by engine according to the discuss in group